### ISSUE TITLE
timepicker: allow two-way binding to something else

### ISSUE THREAD

**Issue by jnizet** [ng-bootstrap/ng-bootstrap#545]
Users of timepicker (and, later, datepicker) will amost certainly want to bind the component to a string, a moment, a Date, or some other class that provides more functionality than a simple object with three fields.

I don't know if the component should provide this functionality (for example by passing a TimeFormat input which would be responsible for transforming the `{hour, minute, second}` object to something else and vice-versa), or if this should be done outside of the component. If the latter, a demo demonstrating how to do that would be helpful.

There might be two different needs BTW: creating a new string every time the model changes, and mutating the time part of a moment/Date.

**Comment by pkozlowski-opensource** [comment]
> Users of timepicker (and, later, datepicker) will amost certainly want to bind the component to a string, a moment, a Date, or some other class that provides more functionality than a simple object with three fields.

Agreed.

> I don't know if the component should provide this functionality

We will never be able to capture all the desired inputs / outputs, that's for sure. But I was under the impression that people could easily do this:

```html
<ngb-timepicker [ngModel]="toStructure(myModel)" (ngModelChange)="fromStructure($event)"
```

where `$event` is the data being updated. Not sure if an additional converter would be less wordy, as it would amount to sth like:

```html
<ngb-timepicker [(ngModel)]="myModel" [timeFormat]="myFormatter"
```

with the downside of needing more code / documentation / explanation on our side.

> There might be two different needs BTW: creating a new string every time the model changes, and mutating the time part of a moment/Date.

The de-sugared `ngModel` handles it without any ambiguity. 

To sum up: I'm not sure what is the best solution in the end but at least I feel like we've got something flexible enough so people can work with any data structure in any way they want. We  can always add more sugar but I would wait for the concrete use-cases before adding it. At least this is how I see it now. Open to discussion.

**Comment by jnizet** [comment]
Good point. I always forget that `[(ngModel)]` can be split into `[ngModel]` and `(ngModelChange)`. I guess this is sufficient for now. Or at least that a `timeFormat` input could be introduced later without preventing us from using only `ngModel` now.

I have the feeling it's more verbose, and that it doesn't encourage reusing code as the second one does, though.

Anyway, are you alright with the idea of adding a demo showing how to write, for example, a custom `StringTimeFormat` class and using it with ngModel to bind to a string? 

I'd be happy to try doing it.

Something like
```
<ngb-timepicker [ngModel]="timeFormat.toStructure(timeAsString)" (ngModelChange)="timeFormat.fromStructure($event)"></ngb-timepicker>
```

and

```
  private timeFormat = new StringTimeFormat();
```

in the component.

**Comment by pkozlowski-opensource** [comment]
Yep, a demo of this would be terrific!

**Comment by jnizet** [comment]
This proves to be harder than what I anticipated. 

Indeed, if the formatter simply returns a new structure every time it parse a string, then the time picker goes into an infinite loop, parsing continuously and making the browser completely unresponsive.

The only way I managed to make it work is to keep a reference to the last parsed structure in memory, and to return that cached instance if the string parses to the same values. But that makes it impossible to use the same timeFormat instance for two different timepicker instances (infinite loop again).

I don't know if that is a design problem of the timepicker, or if that's unavoidable.

For the record, here is the code that works fine, as long as you use the timeFormat for a single timepicker:

View:

```
<ngb-timepicker [ngModel]="timeFormat.toStructure(time)" (ngModelChange)="time = timeFormat.fromStructure($event)"></ngb-timepicker>
<button type="button" class="btn btn-primary" (click)="resetToMidnight()">Reset to midnight</button>
<hr>
<pre>Selected time: {{ time }}</pre>
```

Component:

```
import {Component} from '@angular/core';
import {NGB_TIMEPICKER_DIRECTIVES} from '@ng-bootstrap/ng-bootstrap';

/**
 * Interface for the structure of the native model of NgbTimepicker
 */
interface TimeStructure {
  hour: number;
  minute: number;
  second: number;
}

/**
 * Left-pads single-digit number with 0, to format 6 as '06', for example
 */
function pad(number): string {
  return number < 10 ? `0${number}` : number;
}

/**
 * Tests if two TimeStructure are both falsy, or equal
 */
function equal(t1: TimeStructure, t2: TimeStructure): boolean {
  if (!t1) {
    return !t2;
  }
  return (!t1 && !t2) || (t1 && t2 && t1.hour === t2.hour && t1.minute === t2.minute && t1.second === t2.second);
}

/**
 * Class allowing to transform a TimeStructure to a string, and vice-versa.
 */
class StringTimeFormat {
  private currentValue: TimeStructure;

  toStructure(timeAsString: string): TimeStructure {
    if (!timeAsString) {
      this.currentValue = null;
    }
    else {
      const parts = timeAsString.split(':');
      const newValue = {
        hour: +parts[0],
        minute: +parts[1],
        second: +parts[2],
      };

      if (!equal(this.currentValue, newValue)) {
        this.currentValue = newValue;
      }
    }
    return this.currentValue;
  }

  fromStructure(t: TimeStructure): string {
    return t && `${pad(t.hour)}:${pad(t.minute)}:${pad(t.second)}`;
  }
}

@Component({
  selector: 'ngbd-timepicker-format',
  templateUrl: './timepicker-format.html',
  directives: [NGB_TIMEPICKER_DIRECTIVES]
})
export class NgbdTimepickerFormat {
  time = '00:00:00';
  timeFormat = new StringTimeFormat();

  resetToMidnight() { this.time = '00:00:00'; }
}
```

**Comment by pkozlowski-opensource** [comment]
> This proves to be harder than what I anticipated. Indeed, if the formatter simply returns a new structure every time it parses a string, then the time picker goes into an infinite loop, parsing continuously and making the browser completely unresponsive.

@jnizet this is not what I see on my side. Here is a plunker I've put together based on the timepicker version from master: https://plnkr.co/edit/Q8zV2N4hcaNavUslx97R?p=preview

For me it all works as expected. Let's iterate on this plunk to expose the eventual problem / streamline the solution to put in demo.

**Comment by jnizet** [comment]
I haven't had a chance to try updating the plunkr, but taking your code for the component exhibits the same issue here. But I'm on master, which uses the new forms module (imports are from  '@angular/forms'), whereas your plunkr uses the deprecated forms (imports are from '@angular/common'). I suspect this is where the problem comes from.

**Comment by pkozlowski-opensource** [comment]
>  But I'm on master, which uses the new forms module (imports are from '@angular/forms'), whereas your plunkr uses the deprecated forms (imports are from '@angular/common'). I suspect this is where the problem comes from.

This is very probable! Let's wait for RC5 with the new forms to land and then re-test.

**Comment by jnizet** [comment]
I just re-tested with RC5, and still have the same issue. `toStructure()` is called continuously :-(

**Comment by pkozlowski-opensource** [comment]
Indeed, I can see this being a problem in RC5 / new forms now. I'm not sure what is going on yet, will take it with the core team if I can't figure it out.

**Comment by pkozlowski-opensource** [comment]
Blocked on https://github.com/angular/angular/issues/11097

**Comment by bedag-moo** [comment]
Another possible design is to dependency inject a global service, like you do with the NgbDateParserFormatter (generally, an API will use the same date representation throughout, so configuring this at the application level will usually be enough).

It's also worth noting that angular/angular#11097 only affects object representations. I happen to only need ISO 8601 strings :-)

This issue is not urgent for me though, because I wrap the NgbInputDatePicker anyway, and have now decorated its ControlValueAccessor and Validator to convert to ISO 8601 and back :-)

**Comment by shtaimoortariq** [comment]
I used this in formControl and it worked for me

**Comment by shyamal890** [comment]
Is there a workaround in the meanwhile?

**Comment by realappie** [comment]
@shyamal890 bedag-moo's solution works fine. The docs is a bit lacking on this but you want to provide an injector across the application by adding this object to the providers array of your `ngModule`. 

`{ provide: NgbDateParserFormatter, useClass: NgbDateParser }`

You can also do this on the component where you are using the date picker, although I prefer the module approach.

In my case `NgbDateParser` is a class that implements the [`NgbDateParserFormatter`](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/datepicker/ngb-date-parser-formatter.ts) abstract class. Which the docs does explain [here](https://ng-bootstrap.github.io/#/components/datepicker/api#NgbDateParserFormatter)

In my case, the `NgbDateParser` looks like this and is utilizing a `MomentNgbDatePickerConvertor` class I wrote for moment to `NgbDateStruct`'s and back conversion.

```typescript
export class NgbDateParser implements NgbDateParserFormatter {
  parse(value: string): NgbDateStruct {
    if ( value ) {
      const momentDate = moment(value);

      const ngbDateStruct = MomentNgbDatePickerConvertor.toNgbDateStruct(momentDate);

      return ngbDateStruct;
    }

    return null;
  }
  format( date: NgbDateStruct ): string {
    let formattedMomentDate = '';

    if ( date ) {
      const momentDate = MomentNgbDatePickerConvertor.toMoment(date);

      formattedMomentDate = momentDate.format('DD-MM-YYYY');
    }

    return formattedMomentDate;
  }
}

/** Converts ngb date structs to moment and back */
export class MomentNgbDatePickerConvertor {

  /** Converts an ngDateStruct to a moment */
  public static toMoment(date: NgbDateStruct): moment.Moment {
    const momentObjectParams = {
      y: date.year,
      // This is because ngb date construct starts month at 0 instead of 1
      M: date.month - 1,
      d: date.day
    };

    const momentObject = moment(momentObjectParams);

    return momentObject;
  }

  /** Takes a moment object and returns a ngbootstrap complaint object */
  public static toNgbDateStruct(date: moment.Moment): NgbDateStruct {
    const ngbDateStruct: NgbDateStruct = {
      day: date.date(),
      month: date.month() + 1,
      year: date.year()
    };

    return ngbDateStruct;
  }
};
```

**Comment by DibranMulder** [comment]
@realappie your solutions comes close. However I still use Angular's locale to get the locale id. Then I use that to set the right locale on moment objects. Then the parsing works fine. 

**Register the locale id.**
```typescript
providers: [
   { provide: LOCALE_ID, useValue: 'en-GB',
]
```
**A customer date parser.**
```typescript
import * as moment from 'moment';
import { NgbDateStruct, NgbDateParserFormatter } from '@ng-bootstrap/ng-bootstrap';
import { Inject, LOCALE_ID } from '@angular/core';

export class NgbDateParser implements NgbDateParserFormatter {
    dateFormat: string;

    constructor(@Inject(LOCALE_ID) private localeId: string) {
    }

    /**
     * Parses the entered date.
     * @param value the entered date.
     */
    parse(value: string): NgbDateStruct {
        if (value) {
            const momentDate = moment(value, 'L', this.localeId);
            if (momentDate.isValid()) {
                const ngbDateStruct = MomentNgbDatePickerConvertor.toNgbDateStruct(momentDate);
                return ngbDateStruct;
            }
        }

        return null;
    }

    /**
     * Formats a date into a language specific string.
     * @param date the data from the model.
     */
    format(date: NgbDateStruct): string {
        let formattedMomentDate = '';

        if (date) {
            const momentDate = MomentNgbDatePickerConvertor.toMoment(date);
            formattedMomentDate = momentDate.locale(this.localeId).format('L');
        }

        return formattedMomentDate;
    }
}

/** Converts ngb date structs to moment and back */
export class MomentNgbDatePickerConvertor {

    public static toMoment(date: NgbDateStruct): moment.Moment {
        const momentObjectParams = {
            y: date.year,
            // This is because ngb date construct starts month at 0 instead of 1
            M: date.month - 1,
            d: date.day
        };

        const momentObject = moment(momentObjectParams);

        return momentObject;
    }


    public static toNgbDateStruct(date: moment.Moment): NgbDateStruct {
        const ngbDateStruct: NgbDateStruct = {
            day: date.date(),
            month: date.month() + 1,
            year: date.year()
        };

        return ngbDateStruct;
    }
}
```
**Use it inside your component.**
```typescript
{ provide: NgbDateParserFormatter, useClass: NgbDateParser }
```
